Refactor: plg_kunena_community

Ruud68 opened the pull request:
Pull Request for Issue # . 
 
#### Summary of Changes 

- Refactored the plugin kunena community, this is untested as I do not have the community extension.
- Refactored the plugin kunena comprofiler, also untested as i do not have the CB extension
- Refactored the plugin kunena easyblog also untested as i do not have the Easyblog extension
- Refactored the plugin kunena gravatar, also untested as i do not have the a gravatar
- Refactored the plugin kunena uddeim, also untested as i do not have uddeim component
- Refactored privacy plugin into subscriber plugin
- Refactored quickicon kunena plugin into service provider / subscriber plugin
- Refactored sampledata kunena plugin into service provider / subscriber plugin
- 
please test and report if things are working and if not, please give detailed information regarding what not

#### Testing Instructions

Ruud68 commented:
note for easylog and uddeim, as they use the respective components libraries, the name spacing for instantiating new classes for these components should be validated / checked / corrected

Ruud68 commented:
Done for now :)

Todo:

- plg_finder_kunena
- plg_kunena_easyblog
- plg_kunena_finder

Ruud68 commented:
Tested:

- installation on existing site works
- new installation on Joomla 6 works
- sample data install works
- quickicon works
- gravatar works (quick test)

Plugins Todo:

- plg_finder_kunena
- plg_kunena_easyblog
- plg_kunena_finder

Let's first test if the other plugins are working correct.

xillibit commented:
- With Jomsocial integration : 

Namespace declaration missing in : Community.php
<img width="2405" height="979" alt="image" src="https://github.com/user-attachments/assets/fde8bf39-0f85-4b0a-a0d1-8f80dfe99278" />

- With Community Builder integration : 
<img width="2821" height="1366" alt="image" src="https://github.com/user-attachments/assets/8bf8dbde-a3e7-430f-b72c-1057fb91bf6b" />

- With Uddeim integration : see my comment

xillibit commented on src/plugins/plg_kunena_uddeim/src/Extension/Uddeim.php:
```diff
@@ -0,0 +1,120 @@
+ * Class PlgKunenaUddeIM
+ * @since Kunena
+ */
+class Easyblog extends CMSPlugin implements SubscriberInterface, DatabaseAwareInterface
```
Should be Uddeim

Ruud68 commented:
@xillibit fixed, and while doing so I also ran into the triggered events: onKunenaDisplay and onKunenaPrepare. I have created Events for these and refactored the triggers so they to are future proof.
Although onKunenaPrepare is triggered a lot, there is actually nothing in the repository (including add-on) that is listening this event?

Note also that the event has e.g. $this->topic as 'subject', but also $this->messages, this would not have worked as $this->topic is an object where $this->messages is an array: I ran into this while implementing the KunenaPrepareEvent which expects subject to be an object (just like the Joomla Core contentprepareevent (which it extends)

Both onKunenaDisplay and OnKunenaPrepare need to be tested if they still work. Hard for me as I do not have the means to test this myself (other then will it work on my environment without throwing errors

xillibit commented:
> Although onKunenaPrepare is triggered a lot, there is actually nothing in the repository (including add-on) that is listening this event?

I don't know neither where these events are used, it could be have been used at some point and removed.


~~I can't do nothing on the frontend have this error now, when trying to acces the forum :~~

xillibit commented:
I have removed everything and done a clean install, now by enabling the CB integration plugin : 
<img width="3839" height="1908" alt="image" src="https://github.com/user-attachments/assets/97d11d7c-d4a3-4c8d-aa59-5baa3d125930" />

When enabling Uddeim integration plugin i can see that in backend under tools : 
<img width="3839" height="1728" alt="image" src="https://github.com/user-attachments/assets/3ac49ceb-a500-412e-8860-4f310db1b4cd" />

Same error that above in backend but in frontend with Uddeim integration plugin enabled : 
<img width="1669" height="1872" alt="image" src="https://github.com/user-attachments/assets/1f293902-9ccb-4596-bb35-4ff4926d29e2" />

Ruud68 commented:
Ok, so this was the code:
```
$this->app->triggerEvent('onKunenaPrepare', ['kunena.topic.list', &$this->topic, &$params, 0]);
```
I refactored that into:
```
        $prepareEvent = new KunenaPrepareEvent('onKunenaPrepare', [
            'context' => 'kunena.topic.list',
            'subject' => $this->topic,
            'params'  => $params,
            'page'    => 0
        ]);
        $this->app->getDispatcher()->dispatch('onKunenaPrepare', $prepareEvent);
```

This is now more strict it it generates the error that you see because $this->topic doesn't exist in the function where it is used in the event.

So I commented out for now as this never worked.

xillibit commented:
Now with las changes with CB integration plugin : 
<img width="3031" height="1803" alt="image" src="https://github.com/user-attachments/assets/c419e766-e453-41f2-a8c2-9f72f48db064" />

xillibit commented:
With Uddeim i have : 

Rendering Error in layout Widget/MenuBar: Class "Kunena\Forum\Plugin\Kunena\Uddeim\KunenaPrivateUddeim" not found in C:\Users\floflo\git\joomla-cms-5.4-dev\libraries\vendor\joomla\event\src\Dispatcher.php on line 454
Layout was rendered in C:\Users\floflo\git\joomla-cms-5.4-dev\components\com_kunena\template\aurelia\layouts\page\default.php on line 27

Showed here : 
<img width="1637" height="1125" alt="image" src="https://github.com/user-attachments/assets/d05c3ee7-a36d-48eb-ba94-a74972ba05a1" />

Ruud68 commented:
Uddeim should be fixed, was namespace issue.

The other error is not logical.
in ListDisplay line 157, KunenaParser::prepareContent($content, 'topicList_default'); is called, this in turn calls:
```
$content = HTMLHelper::_('content.prepare', $content, $params, 'kunena.contentprepare.' . $target);
```
So that line triggers onContentPrepare, and not onKunenaPrepare (which is what the error is about).

Can you check in you installation if this is also the case for you?

xillibit commented:
With your last changes i don't have this error with Uddeim, so with Uddeim i don't have any error now

Ruud68 commented:
> Now with las changes with CB integration plugin : <img alt="image" width="2000" height="1803" src="https://private-user-images.githubusercontent.com/876395/489026837-c419e766-e453-41f2-a8c2-9f72f48db064.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTc3Njc5MTQsIm5iZiI6MTc1Nzc2NzYxNCwicGF0aCI6Ii84NzYzOTUvNDg5MDI2ODM3LWM0MTllNzY2LWU0NTMtNDFmMi1hOGMyLTlmNzJmNDhkYjA2NC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwOTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDkxM1QxMjQ2NTRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00M2JiZDc0MmJkOTBlOWFjZjBmMTA0OTA4MDdlYWQzN2NmZjNlM2M5MGNiNGFmOGJiNzQwMGU5Mjk4YjMwZGE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._KZX3Y09QWHEs2BOyHuRd_sT_N5B2CduMqKbugBTocw">

For the check I was referring to this one (comprofiler)

xillibit commented:
I had missed some files now with CB integration i don't have any error

Ruud68 commented:
and the comprofiler 'onKunenaprepare' error is now also gone?

xillibit commented:
> and the comprofiler 'onKunenaprepare' error is now also gone?

yes it's gone

Ruud68 commented:
Ok, so also added plg_kunena_finder as subscriber plugin

xillibit commented:
By comparing with Kunena 6.4.x when i enable CB integration, i don't have the avatar, the link to profle and the private link from CB. On K 6.4 : 
<img width="2951" height="1771" alt="image" src="https://github.com/user-attachments/assets/df46c7ea-f79c-4f54-aa74-e1261bdc9325" />

<img width="1613" height="1059" alt="image" src="https://github.com/user-attachments/assets/ca0111a3-c402-4f76-a89a-f4f6ca2d0289" />


On K6.5 with our PR :  
<img width="2475" height="1652" alt="image" src="https://github.com/user-attachments/assets/19318c1a-ea04-45a1-a133-8d31d32185cd" />

Ruud68 commented:
Hi @xillibit , can you provide me with a backup (akeeba) of your test site, so i can see what it is (not) doing? Without errors this is impossible to troubleshoot.

xillibit commented:
Hi, i have just send to you a mail with a link into it to download a backup

Ruud68 commented:
Hi @xillibit, Found it (I hope). Was not an issue with the code, but with the setting.
an you disable avatar and profile in plugin_kunena_kunena?

Because you have set that as the avatar and profile provider as well, it is the first to respond and it will use that plugins avatar instead of the comprofile one.

Can you check that?

xillibit commented:
By disabling avatar and profile, i get avatar and profile link from CB. But it's not the same behaviour has K6.4, i don't have to disable avatar and profile in plugin_kunena_kunena to get avatar and profile link from CB : 
<img width="2779" height="1002" alt="image" src="https://github.com/user-attachments/assets/86480507-864d-4684-a257-3c45211f38cf" />

Ruud68 commented:
On my production site I have a plugin for fetching the avatar / profile from ochBlog (basically user custom fields).
When I set plg_kunena_kunena avatar = yes, it will get the avatar from Kunena en not from my plugin, setting it back to no fetches the avatar from plg_kunena_ochblog again.

How triggered events work is that the code triggers the event onKunenaGetAvatar.
every plugin that listens to that event will answer, answering will be done in plugin order.

So the result is an array with all the provided avatars in the order of the plugin.

in 6.4 but also in 6.5 we use the first reported avatar in the array.

in 6.4, you can see that it will set the self::$instance to the first $class that is an object (avatar) >
```
            $classes = Factory::getApplication()->triggerEvent('onKunenaGetAvatar');

            foreach ($classes as $class) {
                if (!\is_object($class)) {
                    continue;
                }

                self::$instance = $class;
                break;
            }
```
this is for 6.5: $avatar is either the avatar or the array (when multiple plugins) of avatars in the order of execution, if it is an array (which it is because both kunena and comprofile give an avatar), it will set self::$instance to the first one $avatar[0]
```
            $avatarEvent = new KunenaGetAvatarEvent('onKunenaGetAvatar');
            Factory::getApplication()->getDispatcher()->dispatch('onKunenaGetAvatar', $avatarEvent);

            $avatar = $avatarEvent->getAvatar();

            if ($avatar instanceof KunenaAvatar) {
                self::$instance = $avatar;
            } elseif (\is_array($avatar) && !empty($avatar) && $avatar[0] instanceof KunenaAvatar) {
                self::$instance = $avatar[0];
            } else {
                self::$instance = new self();
            }
```

So I think that on your kunena 6.4 site the plg_kunena_kunena is ordered (or if not ordered, executed first) before the plg_kunena_comprofiler

xillibit commented:
I have tested on K6.4 by changing the order of the plg_kunena_kunena and so you are right when i put in first it doesn't get the CB avatar and when it's in last it get the avatar from CB

Ruud68 commented:
And while investigating that I found two other issues:
we have events that should only return one result: avatar, private, profile
we have events that should return multiple results: accesscontrol, activity, login

So or the single events, I have added a stoppropagation, this to avoid the second triggered plugin overwriting the first triggered plugin value: this then resulted in e.g. the avatar of the last plugin being displayed first.

I have refactored the 'multiple' events so the triggered event now gets an array with all the objects in it: it can then use all or decide which one to use, I crawled through the old code and this is the same now.

So all should be okay now, except for the easysocial plugin as that is not yet converted to a subscriber / service provider plugin.
